File: check_jsx.js

const fs = require('fs');

function checkBalance(filename) {
    const content = fs.readFileSync(filename, 'utf-8');
    const lines = content.split('\n');
    const stack = [];

    // Simple regex: <div ... >, </div>, <img ... />, <br />
    // Group 1: Closing /
    // Group 2: Tag Name
    // Group 3: Self-closing /
    const tagRe = /<(\/)?([a-zA-Z0-9\.]+)(?:[^>]*?(\/)?)?>/g;

    const voidTags = new Set(['br', 'hr', 'img', 'input', 'link', 'meta', 'area', 'base', 'col', 'embed', 'source', 'track', 'wbr']);

    for (let i = 0; i < lines.length; i++) {
        const line = lines[i];
        let match;
        // Reset regex index for each line? No, regex is stateful if global. 
        // But here we create new regex or just match string.
        // Actually, matching line by line is safer for reporting line numbers.

        // We need to be careful about matching tags inside strings.
        // This simple parser will fail on complex cases, but might catch the div issue.

        while ((match = tagRe.exec(line)) !== null) {
            const isClosing = match[1] === '/';
            const tagName = match[2];
            const isSelfClosing = match[3] === '/';

            if (voidTags.has(tagName) || isSelfClosing) {
                continue;
            }

            if (isClosing) {
                if (stack.length === 0) {
                    console.log(`Error: Unexpected closing tag </${tagName}> at line ${i + 1}`);
                    return;
                }

                const lastTag = stack.pop();
                if (lastTag !== tagName) {
                    console.log(`Error: Mismatched closing tag. Expected </${lastTag}> but found </${tagName}> at line ${i + 1}`);
                    // return; // Keep going to see more errors? No, usually cascading.
                }
            } else {
                stack.push(tagName);
            }
        }
    }

    if (stack.length > 0) {
        console.log(`Error: Unclosed tags at EOF: ${JSON.stringify(stack)}`);
    } else {
        console.log("Structure seems balanced.");
    }
}

checkBalance('c:/Aldo/PetMate/petmate-landing-mvp/pages/sitter.tsx');

File: check_jsx.py

import re

def check_balance(filename):
    with open(filename, 'r', encoding='utf-8') as f:
        lines = f.readlines()

    stack = []
    # Simple regex for finding tags. 
    # Captures: 1=Closing slash (/), 2=Tag Name, 3=Self-closing slash (/)
    tag_re = re.compile(r'<(/)?([a-zA-Z0-9\.]+)(?:[^>]*?(/)?)?>')

    for i, line in enumerate(lines):
        # Remove anything inside {...} to avoid confusion with JS logic (simplification)
        # This is flawed for complex JSX but might catch the div issue.
        # Actually, let's just parse tags.
        
        # We need to ignore tags inside strings or comments, but that's hard.
        # Let's try a naive approach first.
        
        matches = tag_re.finditer(line)
        for match in matches:
            is_closing = match.group(1) == '/'
            tag_name = match.group(2)
            is_self_closing = match.group(3) == '/'

            if tag_name in ['br', 'hr', 'img', 'input', 'link', 'meta']: # Void elements
                continue
                
            if is_self_closing:
                continue

            if is_closing:
                if not stack:
                    print(f"Error: Unexpected closing tag </{tag_name}> at line {i+1}")
                    return
                
                last_tag = stack.pop()
                if last_tag != tag_name:
                    print(f"Error: Mismatched closing tag. Expected </{last_tag}> but found </{tag_name}> at line {i+1}")
                    return
            else:
                stack.append(tag_name)

    if stack:
        print(f"Error: Unclosed tags at EOF: {stack}")
    else:
        print("Structure seems balanced.")

check_balance('c:/Aldo/PetMate/petmate-landing-mvp/pages/sitter.tsx')
